streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -476,7 +491,7 @@ public void close() {
      */
     public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
         log.debug("{} Stopping Kafka Stream process.", logPrefix);
-        if (state.isCreatedOrRunning()) {
+        if (state.isCreatedOrRunning() || globalStreamThread != null) {
```

Comment:
The code needs to still run if the global stream thread is running. So all the other threads can be dead, but if the global one is alive, the code needs to run.